File: src/geg/core/state.py
"""Derived election lifecycle state (DESIGN.md §4.2).

Election state is **derived, never stored**: a pure function of the facts in the
data layer and the current time. Every service and every auditor uses this same
function, so no service owns transitions — they are all watchers. The voting
interval is half-open ``[voting_start, voting_end)``: a ballot at exactly
``voting_end`` is out of window.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from geg.core.config import ElectionConfig


class ElectionState(str, Enum):
    # Terminal states.
    CANCELLED = "Cancelled"  # cancellation recorded (only possible before voting_start)
    COMPLETE = "Complete"  # result published
    VOID = "Void"  # now >= tally_deadline and no result
    DKG_FAILED = "DKGFailed"  # now >= voting_start and key not finalized
    # Live states.
    REGISTERED = "Registered"  # config exists, key not finalized, now < voting_start
    KEY_READY = "KeyReady"  # key finalized, now < voting_start
    VOTING = "Voting"  # key finalized, voting_start <= now < voting_end
    TALLYING = "Tallying"  # key finalized, now >= voting_end, no result, now < tally_deadline


@dataclass(frozen=True)
class StateFacts:
    """The data-layer facts state derivation depends on (DESIGN.md §4.2)."""

    cancelled: bool  # a cancellation fact exists
    key_finalized: bool  # the DKG finalization quorum rule is met
    result_published: bool  # a result artifact exists


def derive_state(config: ElectionConfig, facts: StateFacts, now: int) -> ElectionState:
    """Map ``(config, facts, now)`` to a lifecycle state (DESIGN.md §4.2, normative).

    The order below is the normative derivation: terminal outcomes are resolved
    before live states.
    """
    # Terminal: cancellation (only ever recorded before voting_start).
    if facts.cancelled:
        return ElectionState.CANCELLED
    # Terminal: a published result wins over everything else.
    if facts.result_published:
        return ElectionState.COMPLETE
    # Terminal: no result by the tally deadline.
    if now >= config.tally_deadline:
        return ElectionState.VOID
    # Terminal: no finalized key by the time voting opens.
    if now >= config.voting_start and not facts.key_finalized:
        return ElectionState.DKG_FAILED

    # Live states (key finalization + the half-open voting window).
    if not facts.key_finalized:
        # now < voting_start guaranteed here (else DKGFailed above).
        return ElectionState.REGISTERED
    if now < config.voting_start:
        return ElectionState.KEY_READY
    if now < config.voting_end:
        return ElectionState.VOTING
    return ElectionState.TALLYING


def is_voting_open(config: ElectionConfig, now: int) -> bool:
    """Half-open voting-window predicate: ``voting_start <= now < voting_end``."""
    return config.voting_start <= now < config.voting_end

File: frontend/shared/src/state.ts
/** Derived election lifecycle — a faithful TS mirror of `geg.core.state.derive_state`.
 *
 * The public API exposes no derived state, so the dashboard computes it from the
 * config timings + the facts it can read (cancelled, finalized key, result). The
 * voting window is half-open `[votingStart, votingEnd)`. */

export type ElectionState =
  | "Cancelled"
  | "Complete"
  | "DKGFailed"
  | "Registered"
  | "KeyReady"
  | "Voting"
  | "Tallying";

export interface StateTimings {
  votingStart: number;
  votingEnd: number;
}

export interface StateFacts {
  cancelled: boolean;
  keyFinalized: boolean;
  resultPublished: boolean;
}

/** Same normative ordering as `derive_state`: terminal outcomes before live states. */
export function deriveState(c: StateTimings, f: StateFacts, now: number): ElectionState {
  if (f.cancelled) return "Cancelled";
  if (f.resultPublished) return "Complete";
  if (now >= c.votingStart && !f.keyFinalized) return "DKGFailed";
  if (!f.keyFinalized) return "Registered"; // now < votingStart guaranteed here
  if (now < c.votingStart) return "KeyReady";
  if (now < c.votingEnd) return "Voting";
  return "Tallying";
}

export function isVotingOpen(c: StateTimings, now: number): boolean {
  return c.votingStart <= now && now < c.votingEnd;
}

/** Non-terminal states never change again on their own. */
export const TERMINAL_STATES: ReadonlySet<ElectionState> = new Set<ElectionState>([
  "Cancelled",
  "Complete",
  "DKGFailed",
]);

/** Theme badge-color modifier class for a lifecycle state (see theme.css). */
export function stateBadgeClass(s: ElectionState): string {
  const map: Record<ElectionState, string> = {
    Complete: "badge--green",
    Voting: "badge--blue",
    KeyReady: "badge--blue",
    Tallying: "badge--amber",
    Registered: "badge--gray",
    Cancelled: "badge--gray",
    DKGFailed: "badge--red",
  };
  return `badge ${map[s]}`;
}
